### ISSUE TITLE
[BUG] PGAP failed, docker exited with rc = 1

### ISSUE THREAD

**Issue by AnmotCar** [ncbi/pgap#315]
This is my first time trying to install and run pgap. I followed the instructions on the quickstart page.

When I run the pipeline on my own data I get the above error and an enormous log, if I run the example file everything works.
Also after the run fails and I try to run again I get a different error indicating the output directory already exists, even thought I deleted it. It will only work if I change its name... and then it cycles back to problem 1. 

I ran the command with --debug and I got he following:

(base) orseniklab@BU105-seq:/mnt/c/WSL/pgap$ ./pgap.py --debug -n -o BC1_results5 /mnt/c/WSL/pgap/input/BC1.fasta
Output will be placed in: /mnt/c/WSL/pgap/BC1_results5
PGAP version 2024-07-18.build7555 is up to date.
WARNING: memory per CPU core (GiB) is less than the recommended value of 2
PGAP failed, docker exited with rc = 1
Unable to find error in log file.

The CPU error is likely because I am running this in WSL (it get the same line on the example file, but it still works).

I am at a loss. my fasta file has no N's and looks like the input file. I even got rid of the other 2 contigs to make it a single contig file but still no dice. 

I attached the cwltool.log file for both the debug and non debug. 

Im really sorry if im just missing something obviuos.

Thank you for your help in advance. 

Anna
[cwltool_debug.log](https://github.com/user-attachments/files/16531211/cwltool_debug.log)
[cwltool.log](https://github.com/user-attachments/files/16531212/cwltool.log)

**Comment by azat-badretdin** [comment]
Thank you for your report, Anna!

The key of dealing with big log file is to look for the first permanentFail message and then go for errors preceding it

The error in the posted log is

```
"/export/home/gpipe/TeamCity/Agent3/work/427aceaa834ecbb6/ncbi_cxx/src/internal/gpipe/app/cloud/cwl/pgapx_yaml_ctl.cpp", line 251: CPgapxYamlCtlApplication::Run() --- Unknown organism paraburkholdaria strydomiana
     Stack trace:
```

I am afraid the spelling of the organism needs to be exact an organism needs to be registered in NCBI Taxonomy

So you are probably looking for `Paraburkholderia strydomiana` 

Please let me know of this helps

**Comment by AnmotCar** [comment]
Thank you for responding so quickly. I didn't expect that. 

The suggestions did and did not fix the problem... I cant believe a typo caused me such a headache.

Changing the taxonomic name did get the pipeline to run. Though very slowly (I left it to run overnight)

This morning it looks like it finished but I'm missing about half of the files that should be there and I also have a large error log again. But there is an annotated file at least =) 

I have the following files:
annot.faa
annot.gbk
annot.gff
annot.sqn
fastaval.xml
annot_translated_cds.faa

So I seem to be missing 
.fna
cds_from_genomic.fna
annot_with_genomic_fasta.gff
calls.tab
checkm.txt

If that helps to know where things broke down.

**Comment by azat-badretdin** [comment]
> So I seem to be missing

Could you please post `cwltool.log` again? Or maybe check that it does not say `permanentFail` anywhere? 

> Thank you for responding so quickly. I didn't expect that.

Appreciated the compliment. Thank you!

**Comment by AnmotCar** [comment]
I did see permanentFail a few times be it was next to completed so I couldn't make sense of it. The first permanentFail was preceded by SIGABRT. Which I assume is a memory problem? 

I have attached the log. Thank you so much for helping me. 

[cwltool.log](https://github.com/user-attachments/files/16550868/cwltool.log)

**Comment by azat-badretdin** [comment]
The last command in cwltool.log boilerplate (We have also the part added by our scripts for statistics) says

```
WARNING Final process status is permanentFail
```
some things are successful - that does not matter, ALL things need to be successful, so any permanentFail in the log means just that - permanentFail.

Alas, in this case first permanentFail happens during third party application calculation: checkm:

```

+ /root/venv/bin/checkm taxonomy_wf -t 1 -g -x fa class Betaproteobacteria bins-prot/ taxonomy_wf-prot/
Process SyncManager-1:
Traceback (most recent call last):
  File "/opt/python-3.9/lib/python3.9/multiprocessing/process.py", line 315, in _bootstrap
    self.run()
  File "/opt/python-3.9/lib/python3.9/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/opt/python-3.9/lib/python3.9/multiprocessing/managers.py", line 583, in _run_server
    server = cls._Server(registry, address, authkey, serializer)
  File "/opt/python-3.9/lib/python3.9/multiprocessing/managers.py", line 156, in __init__
    self.listener = Listener(address=address, backlog=16)
  File "/opt/python-3.9/lib/python3.9/multiprocessing/connection.py", line 448, in __init__
    self._listener = SocketListener(address, family, backlog)
  File "/opt/python-3.9/lib/python3.9/multiprocessing/connection.py", line 591, in __init__
    self._socket.bind(address)
OSError: [Errno 95] Operation not supported
Traceback (most recent call last):
  File "/root/venv/bin/checkm", line 856, in <module>
    checkmParser.parseOptions(args)
  File "/root/venv/lib/python3.9/site-packages/checkm/main.py", line 992, in parseOptions
    self.analyze(options)
  File "/root/venv/lib/python3.9/site-packages/checkm/main.py", line 326, in analyze
    binIdToModels = mgf.find(binFiles,
  File "/root/venv/lib/python3.9/site-packages/checkm/markerGeneFinder.py", line 68, in find
    binIdToModels = mp.Manager().dict()
  File "/opt/python-3.9/lib/python3.9/multiprocessing/context.py", line 57, in Manager
    m.start()
  File "/opt/python-3.9/lib/python3.9/multiprocessing/managers.py", line 558, in start
    self._address = reader.recv()
  File "/opt/python-3.9/lib/python3.9/multiprocessing/connection.py", line 250, in recv
    buf = self._recv_bytes()
  File "/opt/python-3.9/lib/python3.9/multiprocessing/connection.py", line 414, in _recv_bytes
    buf = self._recv(4)
  File "/opt/python-3.9/lib/python3.9/multiprocessing/connection.py", line 383, in _recv
    raise EOFError
EOFError

```

We had a checkm issue in #296 (JIRA: PGAPX-1223) we were not able to reproduce it

This issue is NOT with a Mac: 

```
{
    "CPU cores": 32,
    "Docker image": "ncbi/pgap:2024-07-18.build7555",
    "cpu flags": "fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl tsc_reliable nonstop_tsc cpuid extd_apicid pni pclmulqdq ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm cmp_legacy cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw topoext ssbd ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 clzero xsaveerptr virt_ssbd arat umip rdpid",
    "cpu model": "AMD Ryzen 9 3950X 16-Core Processor",
    "max user processes": "unlimited",
    "memory (GiB)": 31.3,
    "memory per CPU core (GiB)": 1.0,
    "open files": 1048576,
    "tmp disk space (GiB)": 931.0,
    "virtual memory": "unlimited",
    "work disk space (GiB)": 295.4
}
```

Also note that the error happens in the same place, while trying to connect to a socket inside Python's multiprocessing module. 

Note that this happens very early, while checkm tries to parse options. 

```

checkmParser.parseOptions(args)

```

Please let me know if you are able to run example from our documentation (Quick Notes)

**Comment by AnmotCar** [comment]
Oh I'm sorry, I should have mentioned. The first thing I did was check if the example dataset works. It does. 

I just ran it again to be absolutely sure and it worked again.

**Comment by azat-badretdin** [comment]
Will you be willing to share your input with us? We keep it confidential. If yes, please email it to prokaryote-tools@ncbi.nlm.nih.gov

We will try locally and we will try to reproduce

**Comment by AnmotCar** [comment]
Yes I dont mind sending the file. 

As an aside. My file is 3 contigs in size. When I keep just the first contig (going from roughly 8 Mbp to 3 Mbp) It took a while but it finished.

**Comment by azat-badretdin** [comment]
I also noticed that you have a very small memory per CPU. I wonder if restricting pgap.py to fewer cores with a `./pgap.py --cpus <n>` would help

**Comment by AnmotCar** [comment]
I beleive that was the problem. I think the system just took all available resources and crashed out around CheckM. I restricted WSL to 8 CPU and 16 GB of RAM which seems to have solved the problem, the annotation worked. It took roughly 5 hours to do 8Mbp (single genome) 

Docker is ignoring my input though. I did --cpus 4 but the cwltool.log states that it had access to 8 CPUs. So I had 1.9GB per CPU.
